### PR TITLE
Changing history API fallback to Express middleware

### PR DESCRIPTION
**Pull request by sebdeckers**
The previous Connect middleware rewrites whether or not the file exists. This causes issues with paths containing dots and paths without extensions. The Express middleware is simpler. It is called only if the file is not found. And it can be passed an absolute file path. E.g. `__dirname + '/public/index.html'`

### CODE REVIEW

File: src/tasks/browserSync.js [hunk 1]
```diff
@@ -14,8 +14,15 @@ class BrowserSyncTask {
 			this.options.config.server = _merge({}, this.options.config.server);
 			this.options.config.server.middleware = _merge([], this.options.config.server.middleware);
 
-			var historyApiFallbackConfig = _merge({}, this.options.historyApiFallback);
-			this.options.config.server.middleware.push(historyApiFallback(historyApiFallbackConfig));
+			let historyApiFallbackConfig;
+			if (typeof this.options.historyApiFallback === 'array') {
```
Comment (lookfirst) [inline src/tasks/browserSync.js]: Would rather to use https://lodash.com/docs#isArray

File: src/tasks/browserSync.js [hunk 2]
```diff
@@ -14,8 +14,15 @@ class BrowserSyncTask {
...
+				historyApiFallbackConfig = this.options.historyApiFallback;
+			} else if (typeof this.options.historyApiFallback === 'string') {
```
Comment (lookfirst) [inline src/tasks/browserSync.js]: isString

**Comment by lookfirst** [comment]
This is a pretty rad implementation. Thanks for the pull request.

File: src/tasks/browserSync.js (hunk 1 above)
Comment (sebdeckers) [inline src/tasks/browserSync.js]: Was considering `Array.isArray()`, is that okay with you?

**Comment by sebdeckers** [comment]
Thanks for the lightning-fast feedback, wow. :sweat_smile:

File: src/tasks/browserSync.js (hunk 2 above)
Comment (sebdeckers) [inline src/tasks/browserSync.js]: Okay let me update that with Lodash as well as the Array version for consistency.

File: src/tasks/browserSync.js (hunk 1 above)
Comment (lookfirst) [inline src/tasks/browserSync.js]: Sure! =) ES6 is fine.

**Comment by sebdeckers** [comment]
@lookfirst Okay, pushed the changes with Lodash. Used rebase fixup so it doesn't spam project history.

**Comment by lookfirst** [comment]
Thanks! =)

**Comment by lookfirst** [comment]
published v 2.0.17

**Comment by zivl** [comment]
sorry if I missed the train...
First, great job @cbas :)
But this change breaks backward compatibility with version `2.0.16` (and all other `^2.0.x`). Hence, major version need to be increment and not minor version.

**Comment by lookfirst** [comment]
How is it not backwards compatible? It should work exactly the same way unless @cbas messed up somehow. =)

**Comment by zivl** [comment]
because it is now checks if `historyApiFallback` is string or is an array, otherwise it throws an error. I think this parameter should be optional and not mandatory as it is now. e.g. what if I want to create my own middleware without fallback (or apply my own fallback policy)?
If you agree with me, I'll fix it in the next few hours...

**Comment by lookfirst** [comment]
I agree... the change should have been backwards compatible... sorry about that...

**Comment by sebdeckers** [comment]
My apologies for the trouble. Was trying to keep it backwards compatible.
1. Should add a test case for this, I suppose.
2. Add a default value (`'index.html'`?)

**Comment by zivl** [comment]
@cbas no, don't add default value at the moment. I see in the code that you took care of `historyApiFallback = undefined` and `historyApiFallback = false` cases. 
What would you suggest to maintain backwards compatible with `historyApiFallback = true`? which was very common to the previous implementation with `connect....`

**Comment by sebdeckers** [comment]
> [bripkens/connect-history-api-fallback](https://github.com/bripkens/connect-history-api-fallback/blob/9155f1384f39ed4e66453f04f41c7b64f29c227b/lib/index.js#L69)
> 
> ``` js
> rewriteTarget = options.index || '/index.html'; 
> ```

The connect middleware rewrites `req.url`. That relies on further middleware to actually handle the request and serve the file.

The express middleware handles the request by serving a path using `res.sendFile` (Express-only, not available in Connect). The `sendFile` demand a file path, either absolute file path or relative file path with a base root dir in the options. Is there any basedir option in `gulp-helpers` that can be used as a reasonable default root dir? Then `index.html` would seem like a backwards compatible default for the `true` case.

**Comment by zivl** [comment]
Agreed. But, when I tested it and used `index.html` as default (or as given parameter), it looks like it works in reversed order. The fallback acts as "fall-first". It serves all resources with my `index.html` - such that the fallback acts right away instead of last. 
Have you encountered that?
(btw, shell we move to another issue? this PR is closed)

**Comment by sebdeckers** [comment]
@zivl It's a middleware like any other. As a fallback you will need to `.use()` it _after_ the static middleware, otherwise it indeed simply serves `index.html` or any matched request (accept header contains html, uses GET method).

**Comment by maxlk** [comment]
@cbas The problem is that this middleware is triggered before browserSync server and doesn't let it handle response correctly. See example in [https://github.com/lookfirst/gulp-helpers/issues/22]

**Comment by sebdeckers** [comment]
@maxlk Any idea how to swap the order of the middleware?

**Comment by maxlk** [comment]
@cbas
Not sure if it's possible with BrowserSync. It doesn't provide any hooks for this purpose (see https://github.com/BrowserSync/browser-sync/issues/204).

If BrowserSync itself was a middleware it would be possible, but it is not. Maybe [connect-browser-sync](https://github.com/schmich/connect-browser-sync) may help.

**Pull request merged**